### DISCUSSION TITLE
Simultaneos Node-level Graph-level Classification

### DISCUSSION THREAD

**Discussion by kavehshamsi** [danielegrattarola/spektral#141] [Q&A] [answered]
I am trying to build a model with two branches, one leading to single output classifying graphs, while one being local to the nodes and classifying nodes. 
This paper talks about it:

https://grlearning.github.io/papers/132.pdf

However, the dataset, and loader objects in spektral make this very difficult. The dataset has space for a single array of labels. The loaders, either take the labels as graph-level, or stack them as node-level. Will this feature be added in the future?

In the meantime, train_on_batch applied to the X part of the batches that come out of dataset.load(), and tracking down the indices and getting the Y labels may be a tedious solution :(

Thank you.

**Comment by danielegrattarola** [comment, marked as answer]
Hi, 

The default Datasets and Loaders in Spektral are meant for more "standard" use cases.

A first consideration is that you are not bound to using them **at all** so, for instance, you can write your training loop using `spektral.data.utils.to_disjoint` to help you with creating batches of graphs manually (including the indices).

A second, more interesting consideration is that labels in the default class for Graph/Dataset/Loader are not assumed to be anything, so you have a lot of flexibility. Here's how you can implement a custom loader for your use case.

Import some stuff here so that you can run the code in a notebook if you want:

```py
import numpy as np, scipy.sparse as sp
from spektral.data import Graph, Dataset, Loader, utils
from spektral.layers.ops import sp_matrix_to_sp_tensor
```

First, let's see how to build a graph with the labels that you need. 
Let's start by creating some dummy matrices from which we can build our graph:

```py
n_nodes = 10
n_feats = 5
n_targets = 2

x = np.ones((n_nodes, n_feats))
a = sp.csr_matrix(np.random.randint(0, 2, (n_nodes, n_nodes)))
y_nodes = np.ones((n_nodes, n_targets))
y_graph = np.ones((n_targets)) + 2
```

Now we can create the graph, passing a tuple `(y_nodes, y_graph)` as graph label: 

```py
g = Graph(x=x, a=a, y=(y_nodes, y_graph))

>>> g
Graph(n_nodes=10, n_node_features=5, n_edge_features=None, n_labels=2)
```

You see that the `n_labels` property is not really correct, but we don't care for now. 

Now we can create a dataset with our a-typical graphs: 

```py
class MyDataset(Dataset):
    def read(self):
        return [
            Graph(x=x, a=a, y=(y_nodes, y_graph)) for _ in range(10)
        ]
data = MyDataset()

>>> data
MyDataset(n_graphs=10)
```


Finally, we create a Loader for our dataset. The functionality of a loader is defined by the `collate` method, which takes a batch of `Graph`s and returns the matrices that we must feed to our model.

See the comments to see what this is doing.

```py
class AdvancedLoader(Loader):
    def collate(self, batch):
        # Built-in method to get all matrices separately by type (x_list, a_list, ...)
        packed = self.pack(batch, return_dict=True)
        
        # We deal with node labels manually
        y = packed.pop('y_list')  # This contains tuples [(y_nodes, y_graphs), ...]
        y_nodes, y_graphs = list(zip(*y))
        y_nodes = np.vstack(y_nodes)  # Stack the labels for node prediction
        y_graphs = np.array(y_graphs)  # Keep as-is for graph prediction

        # Convert x_list and a_list to disjoint mode inputs
        inputs = utils.to_disjoint(**packed)

        # Convert any scipy.sparse matrix to SparseTensor
        inputs = list(inputs)
        for i in range(len(inputs)):
            if sp.issparse(inputs[i]):
                inputs[i] = sp_matrix_to_sp_tensor(inputs[i])
        inputs = tuple(inputs)

        # Return the inputs and targets
        return inputs, y_nodes, y_graphs

loader = AdvancedLoader(data, batch_size=3) 
```

So now we can use the Loader in our training loop (I'm not sure whether it will work with `model.fit`):

```py
for batch in loader.load():
    inputs, y_nodes, y_graphs = batch
    model.train_on_batch(inputs, [y_nodes, y_graphs])
```

Note that `train_on_batch` stopped working in TF 2.4, so you'll have to write your own training function if you're on 2.4. There is an example [here](https://github.com/danielegrattarola/spektral/blob/master/examples/graph_prediction/general_gnn.py).

Also, this loader can also be used with edge attributes, if you have them. 

Hope this helps.
Cheers, 
Daniele

**Reply by danielvlla** [reply]
Hi Daniele, can a Loader be used to pass in 2 graphs as inputs to the network - with one resulting label? I am trying to create a Siamese Network where 2 graphs are compared and the y label is either 0 or 1 if the 2 graphs are from the same class or not. I already created the network using GCNs from your library, but I still need to create the Loader to take in inputs as 2 graphs for my dataset.

`inputs=[graph1, graph2], y=[0,1]`

**Reply by danielegrattarola** [reply]
You can definitely do it, you just need to implement a custom loader that returns the tensors that you want. 
The example I posted above is a good starting point for it, let me know if you need help with the implementation.

Cheers
